tests/test_load_balancing_client.py

```diff
@@ -0,0 +1,169 @@
+                [
+                    (
+                        AddressFamily.AF_INET,
+                        SocketKind.SOCK_STREAM,
+                        6,
+                        "",
+                        ("127.0.0.1", port),
+                    ),
+                    (
+                        AddressFamily.AF_INET,
+                        SocketKind.SOCK_DGRAM,
+                        17,
+                        "",
+                        ("127.0.0.1", port),
+                    ),
+                    (
+                        AddressFamily.AF_INET,
+                        SocketKind.SOCK_RAW,
+                        0,
+                        "",
+                        ("127.0.0.1", port),
+                    ),
+                ]
+            )
+        socket_mock.return_value = response_list
+        yield
+
+
+def test_client_works():
+    """Basic test- does it turn on"""
+    with mock_tgis_server(9000):
+        wrapper = GRPCLoadBalancer(client_class=generation_pb2_grpc.GenerationServiceStub, target="localhost:9000")
+        client = wrapper.get_client()
+
+        response = client.Tokenize(request=generation_pb2.BatchedTokenizeRequest())
+        assert response.responses[0].token_count == 5
+
+
+def test_target_validation():
+    """Targets must be in host:port format"""
+    with pytest.raises(ValueError, match="Target must be provided in .* format"):
+        GRPCLoadBalancer(client_class=generation_pb2_grpc.GenerationServiceStub, target="localhost")
+
+    with pytest.raises(ValueError, match="Target must be provided in .* format"):
+        GRPCLoadBalancer(client_class=generation_pb2_grpc.GenerationServiceStub, target="9001")
+
+    with pytest.raises(ValueError, match="Target must be provided in .* format"):
+        # NB: dns targets not supported
+        GRPCLoadBalancer(client_class=generation_pb2_grpc.GenerationServiceStub, target="dns://foo.bar/localhost:9001")
+
+
+def test_client_rebuilds_on_ip_change():
+    """If a new pod in the target service appears, the grpc load balancer won't have any trigger
+    to re-query DNS. Forcing a new client with a new channel will pick up the new pod."""
+
+    poll_interval = 0.0001  # 0.1 ms
+    with mock_tgis_server(9000):
+        with mock_ip_set([8080]):
+            wrapper = GRPCLoadBalancer(client_class=generation_pb2_grpc.GenerationServiceStub,
+                                       target="localhost:9000",
+                                       poll_interval_s=poll_interval)
+            client = wrapper.get_client()
+            client.Tokenize(request=generation_pb2.BatchedTokenizeRequest())
+
+        with mock_ip_set([8080, 9090]):
+            then = datetime.datetime.now()
+            while client is wrapper.get_client():
+                assert datetime.datetime.now() - then < datetime.timedelta(milliseconds=100), "Client did not update"
+
+            # new client still works
+            new_client = wrapper.get_client()
```

Comment:
Hm, maybe I'm wrong about the deadlock? This may indicate that there's an unreachable path since `self._poll_for_ips()` runs synchronously before `__init__` finishes and therefore you can't have `get_client` called while `self._client` is `None`. Then again, if the same client object could start initializing from one thread and then be used from another before initialization has finished, it _could_ happen. Probably very unlikely/impossible though.